File: ios/Ciphio/Ciphio/PasswordGenerator.swift
import Foundation

struct PasswordConfig {
    var length: Int = 16
    var includeUppercase: Bool = true
    var includeLowercase: Bool = true
    var includeDigits: Bool = true
    var includeSymbols: Bool = true
}

struct PasswordStrength {
    let entropyBits: Double
    let label: String
}

final class PasswordGenerator {
    private let uppercase = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
    private let lowercase = Array("abcdefghijklmnopqrstuvwxyz")
    private let digits = Array("0123456789")
    private let symbols = Array("!@#$%^&*()-_=+[]{};:,.<>?/|")

    func generate(config: PasswordConfig) -> String {
        let pool = characterPool(config: config)
        guard !pool.isEmpty else { return "" }
        guard (6...64).contains(config.length) else { return "" }

        var password = String()
        while password.count < config.length {
            var random: UInt32 = 0
            let status = SecRandomCopyBytes(kSecRandomDefault, MemoryLayout<UInt32>.size, &random)
            if status != errSecSuccess { continue }
            let index = Int(random) % pool.count
            password.append(pool[index])
        }
        return password
    }

    func strength(for config: PasswordConfig) -> PasswordStrength {
        let poolSize = Double(characterPool(config: config).count)
        guard poolSize > 0 else { return PasswordStrength(entropyBits: 0, label: "Invalid") }
        let entropy = Double(config.length) * log2(poolSize)
        let label: String
        switch entropy {
        case ..<40: label = "Weak"
        case ..<60: label = "Moderate"
        case ..<100: label = "Strong"
        default: label = "Very Strong"
        }
        return PasswordStrength(entropyBits: entropy, label: label)
    }

    private func characterPool(config: PasswordConfig) -> [Character] {
        var pool = [Character]()
        if config.includeUppercase { pool += uppercase }
        if config.includeLowercase { pool += lowercase }
        if config.includeDigits { pool += digits }
        if config.includeSymbols { pool += symbols }
        return pool
    }
}

File: CiphioVault/CiphioVault/android/app/src/main/java/com/ciphio/vault/password/PasswordGenerator.kt
package com.ciphio.vault.password

import java.security.SecureRandom
import kotlin.math.log2

data class PasswordConfig(
    val length: Int,
    val includeUppercase: Boolean,
    val includeLowercase: Boolean,
    val includeDigits: Boolean,
    val includeSymbols: Boolean
)

data class PasswordStrength(
    val entropyBits: Double,
    val label: String
)

class PasswordGenerator(private val secureRandom: SecureRandom = SecureRandom()) {

    private val uppercase = ('A'..'Z').joinToString("")
    private val lowercase = ('a'..'z').joinToString("")
    private val digits = ('0'..'9').joinToString("")
    private val symbols = "!@#\$%^&*()-_=+[]{};:,.<>?/|"

    fun generate(config: PasswordConfig): String {
        val pool = buildPool(config)
        require(pool.isNotEmpty()) { "At least one character set must be selected" }
        require(config.length in 6..64) { "Length must be between 6 and 64" }

        return buildString {
            repeat(config.length) {
                append(pool[secureRandom.nextInt(pool.length)])
            }
        }
    }

    fun calculateStrength(config: PasswordConfig): PasswordStrength {
        val poolSize = buildPool(config).length
        if (poolSize == 0 || config.length <= 0) {
            return PasswordStrength(0.0, "Invalid")
        }
        val entropy = config.length * log2(poolSize.toDouble())
        val label = when {
            entropy < 40 -> "Weak"
            entropy < 60 -> "Moderate"
            entropy < 100 -> "Strong"
            else -> "Very Strong"
        }
        return PasswordStrength(entropy, label)
    }

    private fun buildPool(config: PasswordConfig): String {
        val pools = buildList {
            if (config.includeUppercase) add(uppercase)
            if (config.includeLowercase) add(lowercase)
            if (config.includeDigits) add(digits)
            if (config.includeSymbols) add(symbols)
        }
        return pools.joinToString(separator = "")
    }
}

